modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3PerProjectClientManager.java

```diff
@@ -0,0 +1,283 @@
+                .entrySet()
+                .stream()
+                // Skip project clients that have no credentials configured. This should not happen in serverless.
+                // But it is safer to skip them and is also a more consistent behaviour with the cases when
+                // project secrets are not present.
+                .filter(entry -> entry.getValue().credentials != null)
+                .collect(Collectors.toUnmodifiableMap(Map.Entry::getKey, Map.Entry::getValue));
+
+            if (clientSettings.isEmpty()) {
+                // clientSettings should not be empty, i.e. there should be at least one client configured.
+                // But if it does somehow happen, log a warning and continue. The project will not have usable client but that is ok.
+                logger.warn("Skipping project [{}] with no client settings", project.id());
+                continue;
+            }
+
+            // TODO: If performance is an issue, we may consider comparing just the relevant project secrets for new or updated clients
+            // and avoid building the clientSettings
+            if (newOrUpdated(project.id(), clientSettings)) {
+                updatedPerProjectClients.put(project.id(), new ClientsHolder(project.id(), clientSettings));
+            }
+        }
+
+        // Updated projects
+        for (var projectId : updatedPerProjectClients.keySet()) {
+            final var old = projectClientsHolders.put(projectId, updatedPerProjectClients.get(projectId));
+            if (old != null) {
+                clientsHoldersToClose.add(old);
+            }
+        }
+        // removed projects
+        for (var projectId : projectClientsHolders.keySet()) {
+            if (currentProjects.containsKey(projectId) == false) {
+                final var removed = projectClientsHolders.remove(projectId);
+                assert removed != null;
+                clientsHoldersToClose.add(removed);
+            }
+        }
+        // Close stale clients asynchronously without blocking the applier thread
+        if (clientsHoldersToClose.isEmpty() == false) {
+            final var currentClientsCloseListener = new SubscribableListener<Void>();
+            final var previousClientsCloseListener = clientsCloseListener;
+            clientsCloseListener = currentClientsCloseListener;
+            if (previousClientsCloseListener != null && previousClientsCloseListener.isDone() == false) {
+                previousClientsCloseListener.addListener(
+                    ActionListener.running(() -> closeClientsAsync(clientsHoldersToClose, currentClientsCloseListener))
+                );
+            } else {
+                closeClientsAsync(clientsHoldersToClose, currentClientsCloseListener);
+            }
+        }
+    }
+
+    private void closeClientsAsync(List<ClientsHolder> clientsHoldersToClose, ActionListener<Void> listener) {
+        executor.execute(() -> {
```

Comment:
Nit: please use an `AbstractRunnable` (here `ActionRunnable.run`) rather than executing a bare `Runnable`, so that exceptions (e.g. a rejection) still complete the listener.